runtime/memprof.c

```diff
@@ -42,54 +565,655 @@ struct memprof_thread_s {
+          memprof_thread_t runner = es->t[i].runner;
+          CAMLassert(runner->running_table == es);
+          CAMLassert(runner->running_index == i);
+          runner->running_index = j;
+        }
+      }
+      ++ j;
+    }
+    ++ i;
+    if (es->young == i) es->young = j;
+    if (es->active == i) es->active = j;
+  }
+  es->evict = es->size = j;
+  CAMLassert(es->active <= es->size);
+  CAMLassert(es->young <= es->size);
+
+  entries_ensure(es, 0);
+}
+
+/* Remove any offset entries from [es]. Ones which have completed an
+ * allocation callback but not a deallocation callback are marked as
+ * deallocated. Others are marked as deleted.
+ *
+ * This is called before moving entries from a thread's entries table
+ * to that of the domain, when we're about to orphan all the domain's
+ * entries. This can occur if we stop a profile and start another one
+ * during an allocation callback (either directly in the callback or
+ * on another thread while the callback is running). We'll never be
+ * able to connect an offset entry to its allocated block (the block
+ * will not be actually allocated until the callback completes, if at
+ * all), but some callbacks may already have been run for it. If no
+ * callbacks have been run, we simply mark the entry as deleted. If
+ * the allocation callback has been run, the best we can do is
+ * probably to fake deallocating the block, so that alloc/dealloc
+ * callback counts correspond.
+ *
+ * Note: no callbacks apart from the allocation callback can run on an
+ * offset entry (as the block has not yet been allocated, it cannot be
+ * promoted or deallocated). */
+
+static void entries_clear_offsets(entries_t es)
+{
+  for (size_t i = 0; i < es->size; ++i) {
+    entry_t e = &es->t[i];
+    if (e->offset) {
+      if (e->callbacks & CB_MASK(CB_ALLOC)) {
+        /* Have called just the allocation callback */
+        CAMLassert(e->callbacks == CB_MASK(CB_ALLOC));
+        e->block = Val_unit;
+        e->offset = false;
+        e->deallocated = true;
+        if (i < es->active) es->active = i;
+      } else {
+        /* Haven't yet called any callbacks */
+        CAMLassert(e->runner == NULL);
+        CAMLassert(e->callbacks == 0);
+        entry_delete(es, i);
+      }
+    }
+  }
+  entries_evict(es);
+}
+
+/* Remove any entries from [es] which are not currently running a
+ * callback. */
+
+static void entries_clear_inactive(entries_t es)
+{
+  CAMLassert (es->config == CONFIG_NONE);
+  for (size_t i = 0; i < es->size; ++i) {
+    if (es->t[i].runner == NULL) {
+      entry_delete(es, i);
+    }
+  }
+  entries_evict(es);
+}
+
+static value validated_config(entries_t es);
+
+/* Transfer all entries from one entries table to another. Return
+ * `false` if allocation fails. */
+
+static bool entries_transfer(entries_t from, entries_t to)
+{
+  size_t delta = to->size;
+
+  if (from->size == 0)
+    return true;
+
+  CAMLassert(validated_config(from) == validated_config(to));
```

Comment:
This assertion does a side effect in the case of a discarded profile. This should generally be avoided, since it may lead to different behavior in debug and normal modes.